x/staking/keeper/msg_server_test.go

```diff
@@ -0,0 +1,1451 @@
+	pubKeys := simapp.CreateTestPubKeys(3)
+
+	validatorAPubKey := pubKeys[1]
+	validatorBPubKey := pubKeys[2]
+
+	delegatorAddress := addresses[0]
+	validatorAAddress := sdk.ValAddress(validatorAPubKey.Address())
+	validatorBAddress := sdk.ValAddress(validatorBPubKey.Address())
+
+	validatorA := teststaking.NewValidator(t, validatorAAddress, validatorAPubKey)
+	validatorB := teststaking.NewValidator(t, validatorBAddress, validatorBPubKey)
+
+	validatorA.Tokens = sdk.NewInt(1_000_000)
+	validatorB.Tokens = sdk.NewInt(1_000_000)
+	validatorA.DelegatorShares = sdk.NewDec(1_000_000)
+	validatorB.DelegatorShares = sdk.NewDec(1_000_000)
+
+	app.StakingKeeper.SetValidator(ctx, validatorA)
+	app.StakingKeeper.SetValidator(ctx, validatorB)
+
+	// The test will go through Delegate/Redelegate/Undelegate messages with the following
+	delegation1Amount := sdk.NewInt(1000)
+	delegation2Amount := sdk.NewInt(1000)
+	redelegateAmount := sdk.NewInt(500)
+	undelegateAmount := sdk.NewInt(500)
+
+	delegate1Coin := sdk.NewCoin(app.StakingKeeper.BondDenom(ctx), delegation1Amount)
+	delegate2Coin := sdk.NewCoin(app.StakingKeeper.BondDenom(ctx), delegation2Amount)
+	redelegateCoin := sdk.NewCoin(app.StakingKeeper.BondDenom(ctx), redelegateAmount)
+	undelegateCoin := sdk.NewCoin(app.StakingKeeper.BondDenom(ctx), undelegateAmount)
+
+	// Delegate to validator A - validator bond shares should not change
+	_, err := msgServer.Delegate(sdk.WrapSDKContext(ctx), &types.MsgDelegate{
+		DelegatorAddress: delegatorAddress.String(),
+		ValidatorAddress: validatorAAddress.String(),
+		Amount:           delegate1Coin,
+	})
+	require.NoError(t, err, "no error expected during first delegation")
+
+	checkValidatorBondShares(validatorAAddress, sdk.ZeroInt())
+	checkValidatorBondShares(validatorBAddress, sdk.ZeroInt())
+
+	// Flag the delegation as a validator bond
+	_, err = msgServer.ValidatorBond(sdk.WrapSDKContext(ctx), &types.MsgValidatorBond{
+		DelegatorAddress: delegatorAddress.String(),
+		ValidatorAddress: validatorAAddress.String(),
+	})
+	require.NoError(t, err, "no error expected during validator bond")
+
+	checkValidatorBondShares(validatorAAddress, delegation1Amount)
+	checkValidatorBondShares(validatorBAddress, sdk.ZeroInt())
+
+	// Delegate more - it should increase the validator bond shares
+	_, err = msgServer.Delegate(sdk.WrapSDKContext(ctx), &types.MsgDelegate{
+		DelegatorAddress: delegatorAddress.String(),
+		ValidatorAddress: validatorAAddress.String(),
+		Amount:           delegate2Coin,
+	})
+	require.NoError(t, err, "no error expected during second delegation")
+
+	checkValidatorBondShares(validatorAAddress, delegation1Amount.Add(delegation2Amount))
+	checkValidatorBondShares(validatorBAddress, sdk.ZeroInt())
+
+	// Redelegate partially from A to B - it should remove the bond shares from the source validator
+	_, err = msgServer.BeginRedelegate(sdk.WrapSDKContext(ctx), &types.MsgBeginRedelegate{
```

Comment:
I was wondering if the delegation was flagged as validator bond for validator B before redelegation.
And then after redelegation, should the validator B's bond shares increase equal to redelegation amount?

Reply:
great catch! Resolved [here](https://github.com/cosmos/cosmos-sdk/pull/16747/commits/846943c2596cf87b9026d3389cd739643b0de453)

Reply:
Awesome 🎉